sql/src/main/java/io/crate/planner/optimizer/rule/RewriteGroupByKeysLimitToTopNDistinct.java

```diff
@@ -0,0 +1,163 @@
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.planner.optimizer.rule;
+
+import io.crate.metadata.TransactionContext;
+import io.crate.planner.operators.GroupHashAggregate;
+import io.crate.planner.operators.Limit;
+import io.crate.planner.operators.LogicalPlan;
+import io.crate.planner.operators.TopNDistinct;
+import io.crate.planner.optimizer.Rule;
+import io.crate.planner.optimizer.matcher.Capture;
+import io.crate.planner.optimizer.matcher.Captures;
+import io.crate.planner.optimizer.matcher.Pattern;
+import io.crate.statistics.TableStats;
+import io.crate.types.DataTypes;
+import org.elasticsearch.Version;
+
+import static io.crate.planner.optimizer.matcher.Pattern.typeOf;
+import static io.crate.planner.optimizer.matcher.Patterns.source;
+
+/**
+ * A rule to rewrite a `SELECT DISTINCT [...] LIMIT n` to use a special TopNDistinct operator.
+ *
+ * <p>
+ * The TopNDistinct operator support early termination, reducing the amount of keys that have to be consumed/processed
+ * from the source.
+ * </p>
+ *
+ * <p>
+ * This optimization is more efficient the lower the limit and the higher the cardinality ratio. (E.g. a unique column).
+ * On a ENUM style column that only has like 5 distinct values in a large table, this TopNDistinct operator could cause a slow down:
+ *
+ * The "early terminate" case wouldn't happen; It would process almost all rows.
+ * In that case our "optimized group by iterator" is more efficient as it contains a ordinal optimization.
+ * (It already knows which values are unique and skips everything else)
+ * </p>
+ */
+public final class RewriteGroupByKeysLimitToTopNDistinct implements Rule<Limit> {
+
+    private final Pattern<Limit> pattern;
+    private final Capture<GroupHashAggregate> groupCapture;
+
+
+    public RewriteGroupByKeysLimitToTopNDistinct() {
+        this.groupCapture = new Capture<>();
+        this.pattern = typeOf(Limit.class)
+            .with(
+                source(),
+                typeOf(GroupHashAggregate.class)
+                    .capturedAs(groupCapture)
+                    .with(groupAggregate -> groupAggregate.aggregates().isEmpty() && eagerTerminateIsLikely(groupAggregate))
+                );
+    }
+
+    private static boolean eagerTerminateIsLikely(GroupHashAggregate groupAggregate) {
+        if (groupAggregate.outputs().size() > 1 || !groupAggregate.outputs().get(0).valueType().equals(DataTypes.STRING)) {
+            // `GroupByOptimizedIterator` can only be used for single text columns.
+            // If that is not the case we can always use TopNDistinct even if a eagerTerminate isn't likely
+            // because a regular GROUP BY would have to do at least the same amount of work in any case.
+            return true;
+        }
+        var cardinalityRatio = groupAggregate.numExpectedRows() / groupAggregate.source().numExpectedRows();
```

Comment:
We should consider the situation that the result of `groupAggregate.source().numExpectedRows()` is zero, otherwise it may throw an `ArithmeticException`.

Reply:
Thank you for catching this.

Reply:
You are welcome.